client.go

```diff
@@ -339,6 +316,54 @@ func (c *Client) Close() error {
 	return nil
 }
 
+// isClosed securely checks if client is marked as closed.
+func (c *Client) isClosed() bool {
+	c.closedMu.RLock(); defer c.closedMu.RUnlock()
+	return c.closed
+}
+
+// setClosed securely marks client as closed (or not closed). If not closed
+// also empty the value inside the startNotify channel by retrieving it (if any),
+// so it doesn't block during connect, when the client was closed and started again,
+// and startNotify was never listened to.
+func (c *Client) setClosed(closed bool) {
+	c.closedMu.Lock(); defer c.closedMu.Unlock()
+	c.closed = closed
+
+	if !closed {
+		// clear channel
+		select {
+		case <-c.startNotify:
+		default:
+		}
+	}
+}
+
+// startNotifyIfNeeded sends ok to startNotify channel if it's listened to.
+// This function is called by connect when connection was successful.
+func (c *Client) startNotifyIfNeeded() {
+	c.closedMu.RLock()
+	if !c.closed {
+		c.log.Debug("sending ok to startNotify chan")
+		select {
+		case c.startNotify <- true:
+		default:
+		// reaching here means the client never read the signal via
+		// StartNotify(). This is OK, we shouldn't except it the consumer
+		// to read from this channel. It's optional, so we just drop the
+		// signal.
+			c.log.Debug("startNotify message was dropped")
+		}
+	}
+	c.closedMu.RUnlock()
+}
+
+// StartNotify returns a channel that receives a single value when the client
+// established a successful connection to the server.
+func (c *Client) StartNotify() <-chan bool {
+	return c.startNotify
```

Comment:
I think this line should be protected with ``c.closedMu.RLock(); defer c.closedMu.RUnlock()`` otherwise you can return the channel just before it should be cleaned in setClosed(false). Shall I create a separate PR for that?